packages/hooks/src/Provider.tsx

```diff
@@ -0,0 +1,94 @@
+import { JsonRpcSigner } from '@ethersproject/providers/src.ts/json-rpc-provider';
+import React from 'react';
+import { ethers } from 'ethers';
+import WalletConnectProvider from '@walletconnect/web3-provider';
+import Web3Modal from 'web3modal';
+
+export interface Web3ContextType {
+  connectWallet: (() => void) | undefined;
+  signer: JsonRpcSigner | null | undefined;
+  userAddress: string | null | undefined;
+  disconnectWallet: (() => void) | undefined;
+  network: string;
+  chainId: number | undefined;
+  connected: boolean;
+  provider: ethers.providers.Web3Provider | null | undefined;
+}
+
+export const Web3Context = React.createContext<Web3ContextType | undefined>(undefined);
```

Comment:
We'll probably want to use a few tricks for defining this context so we (or our users) don't use to check for its existence all over the place before we can use it. 

```ts
// without tricks 
const web3Context = useContext(Web3Context)
// gross
web3Context?.connectWallet()

// with tricks
const web3Context = useContext(Web3Context)
// yum
web3Context.connectWallet()
```

We'll get to it later!

Reply:
Yeah this is a massive pain point.